fathom/src/surface/elaboration/patterns.rs

```diff
@@ -0,0 +1,755 @@
+
+    pub fn row(&self, index: usize) -> &PatRow<'arena> {
+        &self.rows[index]
+    }
+
+    pub fn row_index(&self, index: usize) -> usize {
+        self.indices[index]
+    }
+
+    /// Collect all the `Constructor`s in the `index`th column
+    pub fn column_constructors(&self, index: usize) -> Vec<Constructor<'arena>> {
+        let mut ctors = Vec::with_capacity(self.num_rows());
+        for (pat, _) in self.column(index) {
+            match pat {
+                CheckedPattern::ReportedError(_)
+                | CheckedPattern::Placeholder(_)
+                | CheckedPattern::Binder(_, _) => continue,
+                CheckedPattern::ConstLit(_, r#const) => ctors.push(Constructor::Const(*r#const)),
+                CheckedPattern::RecordLit(_, labels, _) => ctors.push(Constructor::Record(labels)),
+            }
+        }
+        ctors.sort_unstable();
+        ctors.dedup();
+        ctors
+    }
+
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = (&PatRow<'arena>, usize)> {
+        self.rows.iter().zip(self.indices.iter().copied())
+    }
+}
+
+pub type PatPair<'arena> = (CheckedPattern<'arena>, Scrutinee<'arena>);
+
+#[derive(Debug, Clone)]
+pub struct PatRow<'arena> {
+    pairs: Vec<PatPair<'arena>>,
+}
+
+impl<'arena> PatRow<'arena> {
+    pub fn new(pairs: Vec<PatPair<'arena>>) -> Self {
+        Self { pairs }
+    }
+
+    pub fn singleton(pair: PatPair<'arena>) -> Self {
+        Self::new(vec![pair])
+    }
+
+    pub fn tail(&self) -> Self {
+        debug_assert!(!self.is_empty());
+        Self::new(self.pairs[1..].to_vec())
+    }
+
+    pub fn len(&self) -> usize {
+        self.pairs.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.pairs.is_empty()
+    }
+
+    pub fn first(&self) -> Option<&PatPair<'arena>> {
+        self.pairs.first()
+    }
+
+    pub fn all_wildcards(&self) -> bool {
+        self.pairs.iter().all(|(pat, _)| pat.is_wildcard())
+    }
+
+    pub fn split_first(&self) -> Option<(&PatPair<'arena>, Self)> {
+        let (first, rest) = self.pairs.split_first()?;
+        Some((first, Self::new(rest.to_vec())))
+    }
+
+    pub fn append(&mut self, mut other: Self) {
+        self.pairs.append(&mut other.pairs);
+    }
+
+    pub fn patterns(&self) -> impl ExactSizeIterator<Item = &CheckedPattern<'arena>> {
+        self.pairs.iter().map(|(pattern, _)| pattern)
+    }
+}
+
+impl<'arena> CheckedPattern<'arena> {
+    /// Specialise `self` with respect to the constructor `ctor`.
+    pub fn specialize(
+        &self,
+        ctx: &mut elaboration::Context<'_, 'arena>,
+        ctor: &Constructor,
+        scrut: &Scrutinee<'arena>,
+    ) -> Option<PatRow<'arena>> {
+        match self {
+            CheckedPattern::ReportedError(range)
+            | CheckedPattern::Placeholder(range)
+            | CheckedPattern::Binder(range, _) => {
+                let columns =
+                    vec![(CheckedPattern::Placeholder(*range), scrut.clone()); ctor.arity()];
```

Comment:
Why is it ok to duplicate these without updating `scrut`?

Reply:
A name/placeholder/error pattern leaves the scrutinee unchanged: it does not project further into it. It is not strictly necessary to duplicate the column: wildcard patterns could be removed entirely from the matrix (indeed, that is how it is done in [How to compile pattern matching](https://julesjacobs.com/notes/patternmatching/patternmatching.pdf)). But the algorithm as presented in [Compiling pattern matching to good decision trees](https://dl.acm.org/doi/10.1145/1411304.1411311) relies on the matrix always being rectangular: each row must have the same number of columns.